assets/app/view/game/buy_trains.rb

```diff
@@ -129,20 +129,37 @@ def render
       end
 
       def from_depot(depot_trains)
+        trains_from_depot = buyable_trains_in_depot(depot_trains, nil)
+        @corporation.abilities(:train_discount) do |ability|
+          trains_from_depot.concat(buyable_trains_in_depot(depot_trains, ability))
+        end
+        trains_from_depot
+      end
+
+      def buyable_trains_in_depot(depot_trains, ability)
         depot_trains.flat_map do |train|
+          next unless ability.nil? || (ability.trains.include? train.name)
+
           train.variants.sort_by { |_, v| v[:price] }.flat_map do |name, variant|
             price = variant[:price]
+            price -= if ability.absolute
+                       ability.discount
+                     else
+                       price * (ability.discount / 100.0)
+                     end unless ability.nil?
```

Comment:
if ability

Reply:
In the latest version I moved the discount calculation to the train_discount ability class, and rewrote it.